File: ci/teamcity/Delft3D/template/determineProduct.kt
package Delft3D.template

import jetbrains.buildServer.configs.kotlin.*
import jetbrains.buildServer.configs.kotlin.buildSteps.*

object TemplateDetermineProduct : Template({

    name = "Determine product to build and test."
    description = "Determine branch prefix for selecting testbenches to run."

    params {
        param("product", "auto-select")
    }

    steps {
        python {
            name = "Determine product by branch prefix"
            command = script {
                content="""
                    if "%product%" == "auto-select":
                        if "pull" in "%teamcity.build.branch%":
                            source_branch = "%teamcity.pullRequest.source.branch%"
                            if source_branch.startswith("revert-") or source_branch.startswith("dependabot/"):
                                product = "tc"
                            else:
                                product = source_branch.split("/")[0]
                        else:
                            product = "%teamcity.build.branch%".split("/")[0]
                        if "%teamcity.build.branch.is_default%" == "true":
                            product = "all"
                        print(f"##teamcity[setParameter name='product' value='{product}-testbench']")
                        print(f"##teamcity[buildNumber '{product}-testbench: %build.vcs.number%']")
                """.trimIndent()
            }
        }
    }
})

File: ci/teamcity/Delft3D/windows/scripts/determineProduct.py
import sys

product = sys.argv[1]
branch = sys.argv[2]
is_default = sys.argv[3]
vcs_number = sys.argv[4]
pr_source_branch = sys.argv[5] if len(sys.argv) > 5 else ""

if product == "auto-select":
    if "pull" in branch:
        if pr_source_branch.startswith("revert-") or pr_source_branch.startswith("dependabot/"):
            product = "tc"
        else:
            product = pr_source_branch.split("/")[0]
    else:
        product = branch.split("/")[0]
    if is_default == "true":
        product = "all"

    print(f"##teamcity[setParameter name='product' value='{product}-testbench']")
    print(f"##teamcity[buildNumber '{product}-testbench: {vcs_number}']")
